networkx/generators/directed.py

```diff
@@ -267,32 +265,44 @@ def _choose_node(G, distribution, delta, psum):
     if abs(alpha + beta + gamma - 1.0) >= 1e-9:
         raise ValueError("alpha+beta+gamma must equal 1.")
 
-    number_of_edges = G.number_of_edges()
+    if delta_in < 0:
+        raise ValueError("delta_in must be >= 0.")
+
+    if delta_out < 0:
+        raise ValueError("delta_out must be >= 0.")
+
+    # pre-populate degree states
+    vs = sum([count * [idx] for idx, count in G.out_degree()], [])
+    ws = sum([count * [idx] for idx, count in G.in_degree()], [])
```

Comment:
This is a neat trick for "raveling" the degree distributions - very nice!

Reply:
Almost! Better formulated would be that we approach uniform sampling when `len(nodes) * delta` >> `len(degree_distribution)` where `degree_distribution` (sampling `candidates` in the code) are the "raveled" in and out-degree distributions (i.e. a node with degree 7 appears 7 times in that list), so its length is equal to `len(nodes) * avg(degree)`. As you can see the inequality then simplifies to `delta` >> `avg(degree)`.

This should make sense because going back to the algorithm, the probability for a node to be sampled scales with its in or out-degree plus the bias, i.e. `~(d+b)`. In case `b` >> `d`, this effectively becomes `~b`, and since the bias is the same for all nodes this indeed effectively is uniform sampling.